zebra-state/src/service/finalized_state/disk_format/upgrade/add_balance_received.rs

```diff
@@ -0,0 +1,362 @@
+                        start_bound..=TransactionLocation::max_for_height(last_tip_height);
+                    for output in db
+                        .transactions_by_location_range(tx_loc_range)
+                        .flat_map(|(_, tx)| tx.outputs().to_vec())
+                    {
+                        // Return early before the next disk read if the upgrade was cancelled and Zebra is shutting down.
+                        if !matches!(cancel_receiver.try_recv(), Err(TryRecvError::Empty)) {
+                            return Err(CancelFormatChange);
+                        } else {
+                            address_received_map
+                                .entry(output.address(network).expect("should be valid"))
+                                .and_modify(make_modifier(output.value.into()))
+                                .or_insert(output.value.into());
+                        }
+                    }
+
+                    // Return early before the next disk read if the upgrade was cancelled and Zebra is shutting down.
+                    if !matches!(cancel_receiver.try_recv(), Err(TryRecvError::Empty)) {
+                        return Err(CancelFormatChange);
+                    }
+
+                    // If the finalized tip has changed while processing the new outputs in blocks since the last tip height,
+                    // continue to the next iteration of the `address_balances` loop to update the received balances by address map.
+                    if last_tip_height != db.finalized_tip_height().expect("state has blocks") {
+                        continue;
+                    }
+
+                    // Read the address balances from disk to an in-memory map of address balances with updated received balances.
+                    let address_balances = address_received_map
+                        // Read balances in parallel by address.
+                        .par_iter()
+                        .map(|(address, &received)| {
+                            // Return an error to short-circuit the parallel iterator and return early
+                            // if the upgrade was cancelled and Zebra is shutting down.
+                            //
+                            // Note: `.collect::<Result<_, _>>()?` will short-circuit a parallel iterator on the first error.
+                            if !matches!(cancel_receiver.try_recv(), Err(TryRecvError::Empty)) {
+                                return Err(CancelFormatChange);
+                            }
+
+                            // Read the address balance from disk.
+                            let mut balance = db
+                                .address_balance_location(address)
+                                .expect("should have address balances in finalized state");
+
+                            // Update the address balance with the tallied received balance.
+                            *balance.received_mut() = received;
```

Comment:
I'm not fully convinced of the correctness of the flow here.

What if:

- An AddressBalanceReceived (ABR) is read, and its `received` is updated in memory
- A block is comitted, and another field of the same ABR is updated (e.g. `balance`)
- The in-memory ABR is written to the DB, overwriting the updated balance
- Below, it is noticed that a block is committed, and the loop continues, but it won't restore the correct balance AFAICT?

I feel like we can't avoid reading and writing to the same field unless it's inside the same database commit.

Reply:
This explains the values in the failed test logs, I'm thinking I'll add a mechanism for pausing block commits before writing to disk from an in-place upgrade.

Alternatives:
- add a mechanism for blocking the state service initialization at startup on the completion of certain format upgrades
- remove the in-place upgrade and make it a major format upgrade

Reply:
Yes, it seems a blocking format upgrade might be useful here, though a bit annoying for operators.

Another alternative would be to use a new column, but that would add redundancy.

Another option: https://github.com/facebook/rocksdb/wiki/Merge-Operator ?

Reply:
Should be fixed by https://github.com/ZcashFoundation/zebra/pull/9295/commits/3ee3253ff0eed0972f7b3a975017ccb5d4e90145 and https://github.com/ZcashFoundation/zebra/pull/9295/commits/f5f88617025276b02656fcaace57996d3b111b39.